MM-35435 - add multiple SKU options to purchase modal

pablovelezvidal opened the pull request:
#### Summary
This PR adds the multiple sku options support to the purchase modal page. Now customers will be able to subscribe to a different cloud tier directly from the webapp.

#### Ticket Link



#### Screenshots
<img width="1761" alt="Captura de pantalla 2021-05-14 a las 19 09 52" src="https://user-images.githubusercontent.com/10082627/118523695-ae4b7180-b73d-11eb-92bb-dd5017538ced.png">


#### Release Note

```release-note
NONE
```

marianunez commented on actions/cloud.tsx:
```diff
@@ -69,3 +74,14 @@ export function completeStripeAddPaymentMethod(stripe: Stripe, billingDetails: B
     };
 }
+
+export function updateCloudSubscription(productId: string) {
```
To be consistent with the actual `client4` action

```suggestion
export function subscribeCloudSubscription(productId: string) {
```

marianunez commented on components/common/radio_group.tsx:
```diff
@@ -0,0 +1,54 @@
+    isDisabled?: (id: string) => boolean | boolean;
+    onChange(e: React.ChangeEvent<HTMLInputElement>): void;
+}
+const RadioButtonGroup: React.FC<RadioGroupProps> = ({
```
Any tests we can add to this new component?

marianunez commented on components/purchase_modal/purchase.scss:
```diff
@@ -72,6 +72,41 @@
                 min-width: 270px;
 
+                .select-plan {
+                    border-bottom: 1px solid rgba(0, 0, 0, 0.08);
```
Should we be using theme colors here?

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -44,17 +47,36 @@ type Props = {
+        keys.forEach((key) => {
+            if (productId && products[key].id === productId) {
+                selectedProduct = products[key];
+            } else if (!productId && products[key].name.includes('Professional')) {
```
If not found should we just fall back on the first one in the list (which should be Professional) so we are not married to a specific name

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -44,17 +47,36 @@ type Props = {
+                selectedProduct = products[key];
+            }
+        });
+    } else if (keys.length === 1) {
+        selectedProduct = products[keys[0]];
+    }
```
Related to the comment above, we should just do this if no selected product is found above.

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -44,17 +47,36 @@ type Props = {
+    selectedProduct: Product | null | undefined;
+}
+
+function findProductInDictionary(products: Dictionary<Product>, productId?: string): Product | null {
```
Should this be a selector instead?

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -64,26 +86,13 @@ export default class PurchaseModal extends React.PureComponent<Props, State> {
             billingDetails: null,
             cardInputComplete: false,
             processing: false,
+            selectedProduct: findProductInDictionary(props.products!),
```
If we don't expect `props.products` to be null then we should remove it from a possible type above in the type definition

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -112,9 +121,57 @@ export default class PurchaseModal extends React.PureComponent<Props, State> {
+    comparePlan = (
+        <a
+            className='ml-1'
+            href='https://mattermost.com/pricing-cloud/'
```
For compare plans like, please use: https://mattermost.com/pricing-cloud/#pricing-grid-block_5fa2028808529

My bad that I didn't include in the latest link in the Figma

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -112,9 +121,57 @@ export default class PurchaseModal extends React.PureComponent<Props, State> {
+    );
+
+    onPlanSelected = (e: React.ChangeEvent<HTMLInputElement>) => {
+        const selectedPlan = findProductInDictionary(this.props.products!, e.target.value);
```
Same comment as above regarding the use of the non-null assertion operator on `products`

marianunez reviewed (changes requested):
Looking great @pablovelezvidal!

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -36,54 +39,65 @@ let stripePromise: Promise<Stripe | null>;
+        keys.forEach((key) => {
+            if (productId && products[key].id === productId) {
+                selectedProduct = products[key];
+            } else if (!productId) {
+                selectedProduct = products[keys[0]];
+            }
```
This check if no productId, doesn't seem related to this for each loop, should we check this at the beginning and return early?

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -36,54 +39,65 @@ let stripePromise: Promise<Stripe | null>;
+    }
+    const keys = Object.keys(products);
+    let selectedProduct = null;
+    if (keys.length > 1) {
+        // here find the product by the provided id, otherwise return the one with Professional in the name
```
What happens if there if there is no keys? It seems that below it still tries to set the first element in the keys array. If there is no products, then it probably should exit early here too

marianunez commented on components/threading/channel_threads/thread_footer/__snapshots__/thread_footer.test.tsx.snap:
```diff
@@ -712,7 +712,7 @@ exports[`components/threading/channel_threads/thread_footer should report total
                 Object {
                   "formatted": <Memo(SemanticTime)
-                    aria-label="Tue Sep 29 2020 02:30:15 GMT+0000"
+                    aria-label="Tue Sep 29 2020 04:30:15 GMT+0200"
```
Are these changes in this thread snap related to this PR?

marianunez commented on components/admin_console/billing/billing_subscriptions/index.tsx:
```diff
@@ -70,6 +70,9 @@ const BillingSubscriptions: React.FC<Props> = () => {
         daysLeftOnTrial = getRemainingDaysFromFutureTimestamp(subscription.trial_end_at);
     }
 
+    isFreeTrial = true;
+    daysLeftOnTrial = 10;
```
We may need to remove this dev testing lines

pablovelezvidal commented:
/update-branch

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -36,54 +39,69 @@ let stripePromise: Promise<Stripe | null>;
+    if (!productId) {
+        return products[keys[0]];
+    }
+    let selectedProduct = null;
+    if (keys.length > 1) {
+        // here find the product by the provided id, otherwise return the one with Professional in the name
```
Given that we already check for `key.length` above this can probably be removed as well and to simplify even further we can just initialized with the first one and replace in the loop if found. 
```suggestion
    let selectedProduct = products[keys[0]];;
    if (keys.length > 1) {
        // here find the product by the provided id, otherwise return the one with Professional in the name
```

marianunez commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -36,54 +39,69 @@ let stripePromise: Promise<Stripe | null>;
+            }
+        });
+    }
+
+    if (!selectedProduct) {
+        selectedProduct = products[keys[0]];
+    }
```
We may be able to remove this based on the above comment

marianunez reviewed (approved):
Great work @pablovelezvidal!

annelieseklein commented:
Looks great @pablovelezvidal ! Just a few questions/comments:

1. What is the size of the plan selector card on the right hand side? In the design I have 270px, but it looks wider here.
2. The copy below the "Subscribe" button looks a bit darker than what is in the design. I have 72% opacity of the center channel text.
3. The color of the radio buttons here also looks darker than what I have in the design.
4. Here it says "Mattermost Cloud Starter....". In the design I have just "Cloud Starter" but I'm not sure if we should already be updating it to just say "Starter" based on the packaging changes coming June 16? @marianunez

pablovelezvidal commented:
> Looks great @pablovelezvidal ! Just a few questions/comments:
> 
> 1. What is the size of the plan selector card on the right hand side? In the design I have 270px, but it looks wider here.
> 2. The copy below the "Subscribe" button looks a bit darker than what is in the design. I have 72% opacity of the center channel text.
> 3. The color of the radio buttons here also looks darker than what I have in the design.
> 4. Here it says "Mattermost Cloud Starter....". In the design I have just "Cloud Starter" but I'm not sure if we should already be updating it to just say "Starter" based on the packaging changes coming June 16? @marianunez

Thanks for the feedback @annelieseklein, here are my updates:

1. Changed in the latest commit.
2. Changed in the latest commit.
3. Changed in the latest commit. One thing to note, the result is the unique height/width that you can manipulate to a radio button by using CSS. This because radio buttons are native elements and their look and feel is specific to each browser and OS. So, the workaround to get the radio button styled has to stay the same size/height of the original button displayed by the browser. Also, check the attached image on how it will display on Firefox. On this round that is the best we can do to style those radio buttons.
4. thoughts @marianunez ?

Chrome:

<img width="1727" alt="Captura de pantalla 2021-05-24 a las 17 25 33" src="https://user-images.githubusercontent.com/10082627/119374596-0c3a0500-bcba-11eb-9198-1cdc68806b83.png">


Safari:
<img width="1607" alt="Captura de pantalla 2021-05-24 a las 17 26 24" src="https://user-images.githubusercontent.com/10082627/119374471-e6acfb80-bcb9-11eb-977e-6b36ec429c97.png">


Firefox:
<img width="1481" alt="Captura de pantalla 2021-05-24 a las 17 54 23" src="https://user-images.githubusercontent.com/10082627/119374390-cf6e0e00-bcb9-11eb-8a39-3c11d16e5a5f.png">

marianunez commented:
> 4\. Here it says "Mattermost Cloud Starter....". In the design I have just "Cloud Starter" but I'm not sure if we should already be updating it to just say "Starter" based on the packaging changes coming June 16? @marianunez

Yes, this name is coming from the products we define in Stripe. We will adjust the name to say "Cloud Starter" or "Starter" based on the final decision for this screen. This will be done in the backend and not related to this PR. Just let us know what is the final copy @annelieseklein

amyblais commented:
@nickmisasi Kind reminder for PR review.

stevemudie reviewed (approved):
Will test post merge.

amyblais commented:
@pablovelezvidal If this is for the next Cloud release, it will need to be cherry-picked to the ``cloud`` branch.

marianunez commented:
@amyblais This would be for Cloud June 16th release so it will probably make it in the next FF
